### PR TITLE
EOS-15352: cfgen doesn't check facter version

### PR DESCRIPTION
**Pull request by knekrasov**
JIRA issue: [EOS-15352](https://jts.seagate.com/browse/EOS-15352)

Solution: check the facter version programmatically at every node specified in the CDF file.

### CODE REVIEW

**Review by mssawant — APPROVED** [review]

File: cfgen/cfgen [hunk 1]
```diff
@@ -476,9 +476,23 @@ def run_command(hostname: str, *args: str) -> str:
...
 
+def validate_facter(hostname: str) -> None:
+    ver_str = run_command(hostname, 'facter', '--version')
+    # The output may contain commit id like this:
+    # 3.14.8 (commit 4339472f441868ecdae694ffc71e7c8ed0fc24e3)
+    ver_str = ver_str.split(' ')[0]
+
+    if version(ver_str) >= version('3.14.8'):
+        return
+    raise RuntimeError(
```
Comment (suvratjoshi) [inline cfgen/cfgen:487]: Is there any way we can make a single attempt to take corrective action here? 
(Something like, if facter is not installed then install this 3.14,8, else if lower ver is installed then remove it and install this 3.14.8)
Even if this attempt fails then error out with appropriate conclusions?

File: cfgen/cfgen [hunk 2]
```diff
@@ -476,9 +476,23 @@ def run_command(hostname: str, *args: str) -> str:
                                    timeout=15).decode()
 
 
+def validate_facter(hostname: str) -> None:
+    ver_str = run_command(hostname, 'facter', '--version')
+    # The output may contain commit id like this:
+    # 3.14.8 (commit 4339472f441868ecdae694ffc71e7c8ed0fc24e3)
+    ver_str = ver_str.split(' ')[0]
+
+    if version(ver_str) >= version('3.14.8'):
```
Comment (suvratjoshi) [inline cfgen/cfgen:485]: Considering https://github.com/Seagate/cortx-hare/pull/1688/files, do we need to make change in the specfile again for sticking only with 3.14.8 and not using ">=", to avoid any surprises that can come with higher versions of facter?

**Comment by suvratjoshi** [comment]
From a comment in `enrich_cluster_desc`, it looks like it used to work with lower version as well before.
Anyway, lets please correct that comment too, as needed.

File: cfgen/cfgen (hunk 2 above)
Comment (knekrasov) [reply cfgen/cfgen:485]: We know that versions lower than 3.0.0 work incorrectly sometimes when network interfaces use labels (like `ipaddress_enp175s0f1:c2`). In particular, we [did see this problem](https://jts.seagate.com/browse/EOS-15307?focusedCommentId=1988577&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-1988577) with facter 2.4.1.

As of now, we can't stick to `3.14.8` version in the RPM (note that we stick to puppet-agent package which is not guaranteed to ship this version always). So if our code is too strict, the user will have problems in the future that (s)he will not manage to resolve.

That's why I suggest keeping this more relaxed check.

File: cfgen/cfgen [hunk 3]
```diff
@@ -476,9 +476,23 @@ def run_command(hostname: str, *args: str) -> str:
                                    timeout=15).decode()
 
 
+def validate_facter(hostname: str) -> None:
```
[the remark this answers is not in the document]
Comment (knekrasov) [reply cfgen/cfgen:479]: 1. It's a double check. Moreover, Hare may have been installed from sources, so RPM-level requirements don't always apply.
2. I am not sure what other dependencies need to be verified. For instance,
   1. Dhall is installed explicitly by Hare. We use very specific Dhall version but as long as our Dhall files are processed (i.e. the compiler doesn't complain on syntax), we're good.
   2. Python-level dependencies are handled by pip package manager. We already have PR #1688 that locks pythonic dependencies

[the remark this answers is not in the document]
Comment (suvratjoshi) [reply cfgen/cfgen:479]: Thanks for explaining
Consul validation?

File: cfgen/cfgen (hunk 1 above)
Comment (knekrasov) [reply cfgen/cfgen:487]: I don't think that it is worth even thinking in that direction.

Adding so much of decision-making at the level of a small tool has a risk to add implicit states. It is not always good when the tool we're using is TOO smart. E.g. the tool was asked to do something, it sees a problem in the input but then it tries to _guess_ what was actually meant, it makes a decision to interpret the input in some other way and then the user/developer comes to the point when (s)he doesn't understand what exactly had happened and why.  Everybody who has ever worked with scripting languages like PHP or Javascript did see the complexity that comes from the thin air just because the language makes too many attempts to guess what was meant (e.g. in PHP the strings that contain numbers may become equal although not all characters match - just because PHP decides that the developer meant that they are numbers (just forgot to cast the string to float, haha), so it compares them as float values and because of IEEE specifics the numbers were decided to be equal enough.

Long story short, I would suggest to keep the things simple and separate responsibilities. Cfgen's responsibility is either generate the configuration or fail if that is not possible. This patch just adds one more pre-validation that can save hours of consequent debugging. Let's not try to save the whole world. It is not always possible.

File: cfgen/cfgen (hunk 3 above)
[the remark this answers is not in the document]
Comment (knekrasov) [reply cfgen/cfgen:479]: Probably. But that check doesn't need to be performed from cfgen. 

@mssawant What's your opinion?

[the remark this answers is not in the document]
Comment (mssawant) [reply cfgen/cfgen:479]: May be we can add these checks (including facter) as part of build (Makefile) itself instead of cfgen. Let the build itself fail before invoking cfgen.

**Review by vaibhavparatwar — APPROVED** [review]

File: cfgen/cfgen (hunk 3 above)
[the remark this answers is not in the document]
Comment (knekrasov) [reply cfgen/cfgen:479]: make is run at the build machine to create an RPM. That RPM is shipped afterwards to the field. So I don't think it makes sense to check the versions from the Makefile, those checks have nothing to do with the runtime environment anyways.

[the remark this answers is not in the document]
Comment (knekrasov) [reply cfgen/cfgen:479]: @suvratjoshi I think I've recalled one use case when this PR is needed and why RPM-level restriction for facter is probably not sufficient.

The RPM checks are always local. So they apply to the RPM packages installed at this particular machine (i.e. where `rpm -i` or `yum install` is invoked). But cfgen goes to another machines where some older RPM packages can be installed (e.g. because of human mistake). There is no guarantee that the remote machine has the proper facter version installed.

**Review by suvratjoshi — APPROVED** [review]
